ray-on-gke/kuberay-tpu-webhook/main.go

```diff
@@ -88,74 +93,192 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
 	return &rayCluster, nil
 }
 
-func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec) string {
-	numWorkers := int(*workerGroupSpec.Replicas)
+func genDNSHostnames(workerGroupSpec ray.WorkerGroupSpec) (string, error) {
+	replicas := workerGroupSpec.Replicas
+	if replicas == nil {
+		return "", errors.New("workerGroupSpec replicas not set")
+	}
+	numWorkers := int(*replicas)
 	workerGroupName := workerGroupSpec.GroupName
 	serviceName := "tpu-worker-group-svc" // TODO: Kuberay should set DNS name for workergroup
 	hostNames := make([]string, numWorkers)
 	for j := 0; j < numWorkers; j++ {
 		hostNames[j] = fmt.Sprintf("%s-%d.%s", workerGroupName, j, serviceName)
 	}
-	return strings.Join(hostNames, ",")
+	return strings.Join(hostNames, ","), nil
 }
 
 func injectHostnames(hostNames string, workerGroupSpec ray.WorkerGroupSpec, workerGroupIndex int, patches *[]patch) {
-	for j := 0; j < len(workerGroupSpec.Template.Spec.Containers); j++ {
-		container := workerGroupSpec.Template.Spec.Containers[j]
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		klog.Fatalf("Container path not specified")
+	}
+	for j := 0; j < len(containers); j++ {
+		container := containers[j]
 		if containerRequestingTPUs(container) {
-			hostNamesPatch := patch {
-				"op": "add",
-			}
-			path := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
+			subdomainPatch, hostNamesPatch := patch{"op": "add",}, patch{"op": "add",}
+			subdomainPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/subdomain", workerGroupIndex)
+			envPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
 			tpuWorkerHostNames := corev1.EnvVar{
 				Name:  "TPU_WORKER_HOSTNAMES",
 				Value: hostNames,
 			}
+			subdomainPatch["path"] = subdomainPath
+			subdomainPatch["value"] = "tpu-worker-group-svc"
```

Comment:
Better yet, define this as a constant so it's easier to find/replace if needed.

Reply:
Done.

Reply:
Actually thinking more on this - the service name should be dependent on the Kuberay cluster name since it's 1:1 with the cluster. So it should be something like "{kuberay-cluster-name}-tpu-worker-svc". 

The service can be created manually for now in the Terraform templates, but the longer term plan should be to create/delete it along with the RayCluster in the operator.